content/blog/2019-10-04-InstructionScheduling.md

```diff
@@ -0,0 +1,143 @@
+
+There are also carry/borrow dependencies which are definitions or uses of carry/borrow bits, which should be treated similarly to a register since they are another stateful processor resource. They are changed during arithmetic operations where a carry or borrow is used, making them frequently defined but rarely used. Adding them to the dependency dag would be unnecessarily constraining, so the authors placed them in a special subgraph for instructions that uses a carry or borrow.
+
+This dag representation differs from other literature on instruction ordering as they include edges for definitions vs definitions. This is necessary for the final definition of a resource at the end of a basic block (as it could be used by instructions that follow the basic block).
+
+
+#### Selecting an Order: The Static Evaluator
+
+Now, using this dag, any scheduling order following a topological sort will produce an execution indistinguishable from the original order.
+
+Their algorithm travels down the dag from the roots and selects *candidates* ---  instructions whose immediate predecessors have all been scheduled (or root instructions).
+
+When choosing the "best" candidate to schedule, they provide two guidelines:
+ 1. Schedule an instruction that will not interlock with the one just scheduled (if possible).
+ 2. Schedule the instruction that is most likely to cause interlocks with instructions after it. If an instruction may cause interlock, they want to schedule it as early as possible.
+
+Lookaheads would definitely improve scheduling, but that also significantly increases worst-case complexity. Instead, they use three concrete heuristics that evaluate the candidates' static local properties. In order of importance, they are as follows:
+ 1. Whether an instruction interlocks with any of its immediate successors.
+ 2. The number of immediate successors.
+ 3. The length of the longest path from the instruction to the leaves.
+
+These criteria yields instructions which:
+1. May cause interlocks. This is desirable because it allows instructions that are likely to interlock to be scheduled as early as possible as that gives the greatest number of candidates for subsequent instructions.
+2. Uncover the most potential successors, thereby giving greater freedom of future choices.
+3. Balance the progress along the paths of the dag, which ensures a more even number of choices throughout the process.
+
+The steps in this algorithm for dag traversal and scheduling are outlined pretty clearly in the paper and so we will not duplicate them here. The final result is displayed in the table above. The original order had four interlocks (referring to instructions via line number): 3-4, 5-6, 7-8. and 8-9. The reordered version only has one: 8-1. 
+
+
+### Computational Complexity of the Algorithm
+Let the number of instructions in the basic block be n.
```

Comment:
FWIW, it is possible to use inline LaTeX if you want (see the README for this repository). But it seems like you're getting by just fine with hacks. :smiley: